mne/bem.py

```diff
@@ -0,0 +1,321 @@
+        for p in range(u['nfit'] - 1):
+            u['M'][k][p] = u['w'][k] * (np.power(mu[p + 1], k1) - mu1n)
+
+
+def _compute_linear_parameters(mu, u):
+    """Compute the best-fitting linear parameters"""
+    _compose_linear_fitting_data(mu, u)
+    uu, sing, vv = linalg.svd(u['M'], full_matrices=False)
+
+    # Compute the residuals
+    u['resi'] = u['y'].copy()
+
+    vec = np.empty(u['nfit'] - 1)
+    for p in range(u['nfit'] - 1):
+        vec[p] = np.dot(uu[:, p], u['y'])
+        for k in range(u['nterms'] - 1):
+            u['resi'][k] -= uu[k, p] * vec[p]
+        vec[p] = vec[p] / sing[p]
+
+    lambda_ = np.zeros(u['nfit'])
+    for p in range(u['nfit'] - 1):
+        sum_ = 0.
+        for q in range(u['nfit'] - 1):
+            sum_ += vv[q, p] * vec[q]
+        lambda_[p + 1] = sum_
+    lambda_[0] = u['fn'][0] - np.sum(lambda_[1:])
+    rv = np.dot(u['resi'], u['resi']) / np.dot(u['y'], u['y'])
+    return rv, lambda_
+
+
+def _one_step(mu, u):
+    """Evaluate the residual sum of squares fit for one set of mu values"""
+    if np.abs(mu).max() > 1.0:
+        return 1.0
+
+    # Compose the data for the linear fitting, compute SVD, then residuals
+    _compose_linear_fitting_data(mu, u)
+    u['uu'], u['sing'], u['vv'] = linalg.svd(u['M'])
+    u['resi'][:] = u['y'][:]
+    for p in range(u['nfit'] - 1):
+        dot = np.dot(u['uu'][p], u['y'])
+        for k in range(u['nterms'] - 1):
+            u['resi'][k] = u['resi'][k] - u['uu'][p, k] * dot
+
+    # Return their sum of squares
+    return np.dot(u['resi'], u['resi'])
+
+
+def _fwd_eeg_fit_berg_scherg(m, nterms, nfit):
+    """Fit the Berg-Scherg equivalent spherical model dipole parameters"""
+    assert nfit >= 2
+    u = dict(y=np.zeros(nterms-1), resi=np.zeros(nterms-1),
+             nfit=nfit, nterms=nterms, M=np.zeros((nterms - 1, nfit - 1)))
+
+    # (1) Calculate the coefficients of the true expansion
+    u['fn'] = _fwd_eeg_get_multi_sphere_model_coeffs(m, nterms + 1)
+
+    # (2) Calculate the weighting
+    f = (min([layer['rad'] for layer in m['layers']]) /
+         max([layer['rad'] for layer in m['layers']]))
+
+    # correct weighting
+    k = np.arange(1, nterms + 1)
+    u['w'] = np.sqrt((2.0 * k + 1) * (3.0 * k + 1.0) /
+                     k) * np.power(f, (k - 1.0))
+    u['w'][-1] = 0
+
+    # Do the nonlinear minimization, constraining mu to the interval [-1, +1]
+    mu_0 = np.random.RandomState(0).rand(nfit) * f
+    fun = partial(_one_step, u=u)
+    cons = []
+    for ii in range(nfit):
+        for val in [1., -1.]:
+            cons.append({'type': 'ineq',
+                         'fun': lambda x: np.array([val * x[ii] + 1.]),
+                         'jac': lambda x: np.array([0.] * ii + [val] +
+                                                   [0.] * (nfit - ii - 1))})
+    mu = optimize.minimize(fun, mu_0, constraints=cons, method='COBYLA',
+                           tol=1e-2).x
```

Comment:
Not too extensively here. This is just fitting the EEG sphere model, and seems to work fine. I did choose this before extensively testing the dipole-fitting solver, and the best one there ended up being COBYLA so perhaps this isn't a terrible choice.